### ISSUE TITLE
Use zip instead of tar in ldc-build-runtime.d

### ISSUE THREAD

**Issue by andre2007** [ldc-developers/ldc#2303]
Is it possible to use the github release zip archive instead of the tar archive in ldc-build-runtime.d?
Phobos has a module for extracting zip archives. The dependency on an external application (tar) would be superfluous and it would be much easier for windows user. The console parameter ldcSrcDir is not self explaining for beginners.

**Comment by JohanEngelen** [comment]
The github zip archive does not contain submodules, so we have to use our selfmade tar archive. But we could also upload a zip archive if that makes things easier.

**Comment by JohanEngelen** [comment]
@andre2007 Do you have an idea for an improved documentation tetx for ldcSrcDir?

**Comment by andre2007** [comment]
Providing a zip file would make the use of ldc-build-runtime.d straight forward on windows too, as you do not need additional applications (tar) or think about an additional command parameter.

For your other questions, I can tell you what I was thinking while trying to start the ldc-build-runtime.d
(I read the WIKI before).

- As described the tool does not work out of the box on windows
- I did a quick search on google whether there is an easy way to install tar on windows (self contained executable does not exist, only a larger package, but it was only a very quick look)
- There is a command parameter ldcSrcDir: Do I have to specify some folder of my ldc installation...?
- After reading the source code of ldc-build-runtime.d, I think I have to download from github releases the right source tar, but there are at least two tars, (the one created by github and the one created by your tools)

**Comment by andre2007** [comment]
There is a tar application available here "http://gnuwin32.sourceforge.net/packages/gtar.htm" but it is not usuable:

 ["tar", "-xzf", "ldc-src.tar.gz", "--strip", "1", "-C", "ldc-src"]
tar: unrecognized option `--strip'
`tar --help' zeigt weitere Informationen.
Error: command failed with status 2

**Comment by kinke** [comment]
[FWIW, `tar.exe` is part of git for Windows, I'm using that one. It's not standalone though, it requires some MSYS DLLs.]

**Comment by joakim-noah** [comment]
I'd like to have a dub file for druntime and Phobos, which could be useful upstream too, but that means reproducing the relevant portions of the current CMake configuration in dub.  This was the easiest way to get going, but we'll get it on dub one day.  @kinke, why not provide a zip file and just use that on Windows with `ldc-build-runtime`?  Seems like using tar is a pain there.

**Comment by kinke** [comment]
> @kinke, why not provide a zip file and just use that on Windows with ldc-build-runtime? Seems like using tar is a pain there.

I don't do everything around here, even my spare time is limited and prioritized. Having had to convert the bash script to D for the sake of the Windows folks (me included) was already a big step towards Windows-friendliness. I guess most interested guys have git installed, that's why using the (ugly - no streaming, no convenience function to extract a full archive) Phobos zlib wrapper (and the inconvenience of having to upload a second source archive) isn't high on my priorities list. @andre2007: PRs are always welcome.

**Comment by joakim-noah** [comment]
Crap, spent five mins. laboriously typing out an answer on my phone, only for there not to be a Preview button on the mobile input box, so hit "Select All" twice, then "Request Desktop Site," and my comment is gone, not in paste either.  Did this damn browser lose my reply?  No, forgot to hit Copy after Select All... gah!!!  Here goes again:

>I don't do everything around here, even my spare time is limited and prioritized.

Sure, I asked because I know you're on Windows, which I haven't used in two years, so it'd be easy for you to test.

@andre2007, you can look at modifying [the packaging scripts to generate a src.zip here](https://github.com/ldc-developers/ldc-scripts/blob/master/ldc2-packaging/3-package-ldc.sh#L36) and easily have [`ldc-build-runtime ` use the zip file on Windows instead](https://github.com/ldc-developers/ldc/blob/master/runtime/ldc-build-runtime.d.in#L140).

**Comment by andre2007** [comment]
I will try to get it working with tar from git bash and if this works I will check whether I can add the zip code into the ldc-build-runtime d coding.
My issue is, I do not know anything about make, cmake, ninja. I only know I want to cross compile to raspberry pi from windows and I have to use ldc-build-runtime. But how to call it, I do not know (Windows user, with no c / c++ experience)
At the moment there are errors thrown while executing from git bash, but I have to analyze deeper.

**Comment by JohanEngelen** [comment]
I've added a zip for 1.4.0-beta1

**Comment by kinke** [comment]
As quick remedy for the tar issue: If bash is an issue, simply add `<git install dir>\usr\bin` (that's the one containing tar.exe in my case for PortableGit) to your PATH environment variable and work as usual in cmd.exe.
Another option is to download the source archive manually (or use the downloaded one in the temporary build directory `ldc-build-runtime.tmp` - either way, the full URL is printed to stdout) and use your unpacker of choice (7zip?) to extract it. Then specify the path to the extracted directory as `--ldcSrcDir`.

Are you going to use the Android NDK?

**Comment by andre2007** [comment]
My intention was to write a blog how to cross compile for Raspberry Pi for beginners (windows). After that I also want to write a similar blog for Android. The more tools you have to install, the more (newbie) users refused to try it out. Therefore I look for opportunities to reduce the complexity for cross compiling.

I will try to create a pull request for extracting the  zip archive from the script.

**Comment by andre2007** [comment]
@JohanEngelen 
Unfortunately the zip is not usable, or I need to write special logic. I need to detect the root folder in the zip file (ldc-1.4.0-beta1-src/). In a default github zip file there is only 1 root folder. But in your zip file there is a second root folder __MACOSX. I could also detect this case, but I think this special logic should not be included..

**Comment by andre2007** [comment]
This should work fine with the corrected zip (only 1 root dir). Could you include this coding?

```
import std.file;
import std.path;
import std.stdio;

enum ldcSourceDir = `C:\Users\user\Desktop\test\ldc-src`;

void main()
{
	extractZipArchive("ldc-1.4.0-beta1-src.zip", ldcSourceDir);
}

void extractZipArchive(string archivePath, string destination)
{
	import std.array: split;
	import std.zip;
	import std.string: endsWith;
	
	ZipArchive archive = new ZipArchive(std.file.read(archivePath));
	string rootDir;
	foreach (name, am; archive.directory)
	{
		auto arr = name.split("/");
		if (arr.length == 2 && arr[1] == "")
		{
			rootDir = name;
			break;
		}
	}

	foreach (name, am; archive.directory)
	{
		if (name == rootDir)
			continue;
		
		string destPath = buildNormalizedPath(destination, name[rootDir.length..$]);

		if(exists(destPath.dirName) == false)
			mkdirRecurse(destPath.dirName);
		
		if(name.endsWith("/"))
			continue;
			
		archive.expand(am);
		std.file.write(destPath, am.expandedData);
	}
}
```

**Comment by JohanEngelen** [comment]
omg :( https://stackoverflow.com/questions/10924236/mac-zip-compress-without-macosx-folder

**Comment by JohanEngelen** [comment]
@andre2007 I've uploaded a new zip.

(future reference: `zip ldc-1.4.0-beta1-src.zip -x \*.DS_Store -r ldc-1.4.0-beta1-src`  )

**Comment by kinke** [comment]
Thanks a lot André, I'll test it and incorporate it then.

**Comment by kinke** [comment]
Implemented in #2314.
